src/translations.h

```diff
@@ -50,16 +59,74 @@
 #  define ATTRIBUTE_FORMAT_ARG(a)
 #endif
 
-const char *_( const char *msg ) ATTRIBUTE_FORMAT_ARG( 1 );
-inline const char *_( const char *msg )
+namespace details
+{
+// same as _(), but without local cache
+const char *_translate_internal( const char *msg ) ATTRIBUTE_FORMAT_ARG( 1 );
+
+inline const char *_translate_internal( const char *msg )
 {
     return msg[0] == '\0' ? msg : gettext( msg );
 }
-inline std::string _( const std::string &msg )
+
+// same as _(), but without local cache
+inline std::string _translate_internal( const std::string &msg )
 {
-    return _( msg.c_str() );
+    return _translate_internal( msg.c_str() );
 }
 
+template<typename T>
+class local_translation_cache;
+
+template<>
+class local_translation_cache<std::string>
+{
+    private:
+        int cached_lang_version = INVALID_LANGUAGE_VERSION;
+        std::string cached_arg;
+        std::string cached_translation;
+    public:
+        const std::string &operator()( const std::string &arg ) {
+            if( cached_lang_version != get_current_language_version() || cached_arg != arg ) {
+                cached_lang_version = get_current_language_version();
+                cached_arg = arg;
+                cached_translation = _translate_internal( arg );
+            }
+            return cached_translation;
+        }
+};
+
+template<>
+class local_translation_cache<const char *>
+{
+    private:
+        std::string cached_arg;
+        int cached_lang_version = INVALID_LANGUAGE_VERSION;
+        bool same_as_arg;
+        const char *cached_translation;
+    public:
+        const char *operator()( const char *arg ) {
+            if( cached_lang_version != get_current_language_version() || cached_arg != arg ) {
+                cached_lang_version = get_current_language_version();
+                cached_translation = _translate_internal( arg );
+                same_as_arg = cached_translation == arg;
+                cached_arg = arg;
+            }
+            // mimic gettext() behavior: return `arg` if no translation is found
+            // `same_as_arg` is needed to ensure that the current `arg` is returned (not a cached one)
+            return same_as_arg ? arg : cached_translation;
+        }
+};
+
+} // namespace details
+
+// Note: in case of std::string argument, the result is copied, this is intended (for safety)
+#define _( msg ) \
+    ( ( []( const auto & arg ) -> auto { \
```

Comment:
```suggestion
    ( ( []( const auto & arg ) { \
```
The `-> auto` is redundant, I think, but you can leave it if you think it adds clarity.

Reply:
You're right, I believe the comment is sufficient.